_posts/2025-01-03-Clang-17-now-in-dev-container.markdown

```diff
@@ -0,0 +1,102 @@
+In all other circumstances, the existing non-temporal version should be preferred for better optimization.
+An example would be using a double-checked pattern when pinning with `heap_claim_fast`, where combining the two valid bit reads would yield incorrect code.
+```c++
+void func(Timeout *t, int *ptr) {
+    if (__builtin_cheri_tag_get_temporal(ptr)) {
+        int claim = heap_claim_fast(t, ptr, nullptr);
+        if (claim == 0 && __builtin_cheri_tag_get_temporal(ptr)) {
+            *ptr = 1234;
+            // ...
+        }
+    }
+}
+```
+
+### Bugfixes
+
+- [Fixed](https://github.com/CHERIoT-Platform/llvm-project/commit/60b4a582dfc1579b3c08c65d4b6ede961eb267f5) a recurring issue where the compiler would generate improperly mangled calls to `memcpy`, `memmove` and/or `memcmp` in specific circumstances, resulting in linker errors. 
+This has now been fixed at the source.
+- [Fixed](https://github.com/CHERIoT-Platform/llvm-project/issues/57) linker errors that arise when taking the address of non-exported, non-libcall functions with non-default interrupt state annotations.
+
+
+### Optimizations
+- [Taught](https://github.com/CHERIoT-Platform/llvm-project/commit/25ad11d7832237e81ca476d4e3e6bac2defc3fa7) the compiler to better optimize `CAndPerm` instructions, including constant folding and idempotence.
+This tends to benefit places where redundant `CAndPerm` instructions were generated by macros or C++ templates.
+- [Freed](https://github.com/CHERIoT-Platform/llvm-project/commit/8221b74cffbfa03149eb5bab1776280ebb43785f) up the TP/X4 register for the compiler's use in code generation.
+This register is normally reserved as a "thread pointer" in RISC-V ABIs, but is not used for that purpose on CHERIoT.
+We haven't observed this making a significant performance or size difference, but some compute-intensive code may benefit.
+
+## Looking Forward
+
+We have a number of further improvements that we expect to make available to CHERIoT toolchain users in the near future:
+
+- [Supporting sealed capabilities](https://github.com/CHERIoT-Platform/llvm-project/pull/88) in the C type system.
+This change adds a new pointer attribute `__sealed_capability`, which disallows any operations that would cause the pointer to be dereferenced, or to be lose its `__sealed_capability` annotation.
+Once integrated with CHERIoT RTOS, we will be able to represent sealing, unsealing, and the propagation of sealed capabilities in a type-safe manner.
+This is demonstrated below.
+```c++
+// CHERI sealing and unsealing operations now have signatures that are type-safe with respect to sealing.
+void * __sealed_capability cheri_seal(void *cap, const void *type);
+void * cheri_unseal(void * __sealed_capability cap, void *type);
+
+int func(int * __sealed_capability ptr) {
+    // This causes a compiler error!
+    return *p;
+}
```

Comment:
The important thing here is that we don’t lose the type info when we seal a T*. You have a sealed T*, you unseal it and get a T*. Today, you have a SObj* and have to know that it should unseal as a T*.

Reply:
We'd need to write a templated version of `cheri_unseal` first, right? We don't have that yet.

Reply:
I don't think we should talk about that in this pre-announcement if we haven't worked it out yet. I could see doing a separate more full announcement once they're completely in and supported in the RTOS.